test/jdk/jdk/jfr/event/gc/detailed/TestGCLockerEvent.java

```diff
@@ -0,0 +1,126 @@
+ * @key jfr
+ * @requires vm.hasJFR
+ * @library /test/lib
+ * @build sun.hotspot.WhiteBox
+ * @run driver ClassFileInstaller sun.hotspot.WhiteBox
+ * @run main/othervm -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI -Xmx32m -Xms32m -Xmn12m -XX:+UseG1GC jdk.jfr.event.gc.detailed.TestGCLockerEvent
+ */
+public class TestGCLockerEvent {
+
+    private static final String EVENT_NAME = EventNames.GCLocker;
+
+    private static final int CRITICAL_THREAD_COUNT = 4;
+
+    private static final CountDownLatch LOCK_COUNT_SIGNAL = new CountDownLatch(CRITICAL_THREAD_COUNT);
+
+    private static final CountDownLatch UNLOCK_SIGNAL = new CountDownLatch(1);
+
+    private static final CountDownLatch UNLOCK_COUNT_SIGNAL = new CountDownLatch(CRITICAL_THREAD_COUNT);
+
+    private static final String CRITICAL_THREAD_NAME_PREFIX = "Critical Thread ";
+
+    private static final int STALL_THREAD_COUNT = 8;
+
+    private static final CountDownLatch STALL_COUNT_SIGNAL = new CountDownLatch(STALL_THREAD_COUNT);
+
+    private static final int LOOP = 32;
+
+    private static final int M = 1024 * 1024;
+
+    public static void main(String[] args) throws Exception {
+        var recording = new Recording();
+        recording.enable(EVENT_NAME);
+        recording.start();
+
+        startCriticalThreads();
+        LOCK_COUNT_SIGNAL.await();
+        startStallThreads();
+        STALL_COUNT_SIGNAL.await();
+
+        // Wait threads to be stalled
+        Thread.sleep(1500);
+
+        UNLOCK_SIGNAL.countDown();
+        UNLOCK_COUNT_SIGNAL.await();
+        recording.stop();
+
+        // Verify recording
+        var all = Events.fromRecording(recording);
+        Events.hasEvents(all);
+        var event = all.get(0);
+
+        assertTrue(Events.isEventType(event, EVENT_NAME));
+        Events.assertField(event, "lockCount").equal(CRITICAL_THREAD_COUNT);
+        Events.assertField(event, "stallCount").atLeast(STALL_THREAD_COUNT);
+        assertTrue(event.getThread().getJavaName().startsWith(CRITICAL_THREAD_NAME_PREFIX));
+
+        recording.close();
+    }
+
+    private static void startCriticalThreads() {
+        for (var i = 0; i < CRITICAL_THREAD_COUNT; i++) {
+            new Thread(() -> {
+                try {
+                    WhiteBox.getWhiteBox().lockCritical();
+                    LOCK_COUNT_SIGNAL.countDown();
+
+                    UNLOCK_SIGNAL.await();
+                    WhiteBox.getWhiteBox().unlockCritical();
+                    UNLOCK_COUNT_SIGNAL.countDown();
+                } catch (InterruptedException ex) {
+                }
+            }, CRITICAL_THREAD_NAME_PREFIX + i).start();
+        }
+    }
+
+    private static void startStallThreads() {
+        var ts = new Thread[STALL_THREAD_COUNT];
+        for (var i = 0; i < STALL_THREAD_COUNT; i++) {
+            ts[i] = new Thread(() -> {
+                STALL_COUNT_SIGNAL.countDown();
+                for (int j = 0; j < LOOP; j++) {
```

Comment:
Since the test already uses WhiteBox, please use whitebox to trigger a gc instead of this dodgy method.

Reply:
Triggering a GC is not enough, I hope these threads could be stall by the GC locker(call GCLocker::stall_until_clear) so that a correct assertion of the number of stall count could be added.
I think it could not be done by WhiteBox::youngGC/fullGC, please correct me if I'm wrong.

Reply:
I agree, although this is more a theoretical concern since it's not checked. Let's keep this for now as is though.